lib/src/async_cache.dart

```diff
@@ -43,70 +48,76 @@ class AsyncCache<T> {
       : _getNow = now;
 
   /// Returns a cached value or runs [callback] to compute a new one.
   ///
   /// If [callback] has been run recently enough, returns its previous return
   /// value. Otherwise, runs [callback] and returns its new return value.
   Future<T> fetch(Future<T> callback()) async {
+    if (_cachedStreamSplitter != null) {
+      throw new StateError('Previously used to cache via `fetchStream`');
+    }
     _invalidateWhenStale();
-    var result = await (_cachedValueFuture ??= callback());
+    if (_cachedValueFuture == null) {
+      _cachedValueFuture = callback();
+    }
+    var result = await _cachedValueFuture;
     _cachedExpiration ??= _getNow().add(_duration);
     return result;
   }
 
   /// Returns a cached stream or runs [callback] to compute a new stream.
   ///
-  /// If [callback] has been run recently enough, returns its previous return
-  /// value. Otherwise, runs [callback] and returns its new return value.
+  /// If [callback] has been run recently enough, returns a copy of its previous
+  /// return value. Otherwise, runs [callback] and returns its new return value.
   ///
   /// If a stream is currently being fetched, waits until the _done_ event and
   /// then returns the cached value.
-  Stream<T> fetchStream(Stream<T> callback()) async* {
-    _invalidateWhenStale();
-    if (_fetchingStream != null) {
-      await _fetchingStream;
+  Stream<T> fetchStream(Stream<T> callback()) {
+    if (_cachedValueFuture != null) {
+      throw new StateError('Previously used to cache via `fetch`');
     }
-    if (_cachedStreamValues != null) {
-      yield* new Stream<T>.fromIterable(_cachedStreamValues);
-      return;
+    if (_isInvalidated) {
+      _cachedStreamSplitter.close();
+      invalidate();
     }
-    var fetching = new Completer();
-    _fetchingStream = fetching.future;
-    var values = <T>[];
-    await for (var result in callback()) {
-      values.add(result);
-      yield result;
+    if (_cachedStreamSplitter == null) {
+      _cachedStreamSplitter = new StreamSplitter<T>(callback());
+      _cachedStreamSplitter.split().last.then((_) {
+        _cachedExpiration = _getNow().add(_duration);
+      });
     }
-    _cachedStreamValues = values;
-    _cachedExpiration ??= _getNow().add(_duration);
-    fetching.complete();
+    return _cachedStreamSplitter.split();
   }
 
   /// Removes any cached value, and returns `true` if a cached value existed.
   Future<bool> invalidate() {
-    if (_cachedExpiration == null) {
-      return new Future<bool>.value(false);
-    }
+    var wasCached = _cachedExpiration == null;
     _cachedExpiration = null;
     _cachedValueFuture = null;
-    _cachedStreamValues = null;
-    return new Future<bool>.value(true);
+    _cachedStreamSplitter = null;
+    return new Future<bool>.value(wasCached);
+  }
+
+  bool get _isInvalidated {
+    return _cachedExpiration != null && !_cachedExpiration.isAfter(_getNow());
   }
 
   void _invalidateWhenStale() {
```

Comment:
It feels strange that we're lazily invalidating the cache on access, rather than eagerly doing it with a timer. I worry that we'll end up keeping large amounts of data in memory uselessly.